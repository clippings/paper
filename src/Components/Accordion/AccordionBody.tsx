import React from 'react';
import { createClassNameModifier } from '../../utils/classNames';
const classNames = {
  body: createClassNameModifier('accordion', 'body'),
};
const AccordionBody: React.FunctionComponent<{}> = ({ children }) => (
  <div className={classNames.body}>{children}</div>
);
export default AccordionBody;
