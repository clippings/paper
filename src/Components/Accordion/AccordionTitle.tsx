import React from 'react';
import classnames from 'classnames';
import { Icon, ICONS, ICONS_COLORS, ICONS_DIRECTIONS } from '../Icon';
import { createClassNameModifier } from '../../utils/classNames';
const titleClassNames = {
  title: createClassNameModifier('accordion', 'title'),
  titleWrap: createClassNameModifier('accordion', 'title__wrap'),
  titleActive: createClassNameModifier('accordion', 'title__active'),
};
type AccordionTitleProps = {
  onClick?: (...args: any[]) => any;
  classNames?: string;
};
const AccordionTitle: React.FunctionComponent<AccordionTitleProps> = ({
  children,
  onClick,
  classNames = '',
  ...rest
}) => (
  <div
    role="presentation"
    className={classnames(titleClassNames.title, classNames)}
    onClick={onClick && onClick}
    {...rest}
  >
    <div className={titleClassNames.titleWrap}>{children}</div>
    <Icon icon={ICONS.CHEVRON} direction={ICONS_DIRECTIONS.DOWN} color={ICONS_COLORS.DARK} />
  </div>
);

export default AccordionTitle;
