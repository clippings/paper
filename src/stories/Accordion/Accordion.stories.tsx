import React from 'react';
import { storiesOf } from '@storybook/react';
import { jsxDecorator } from 'storybook-addon-jsx';
import { Accordion } from '@src';

const items = [
  {
    id: 1,
    title: 'Item title',
    body: 'Item body',
  },
  {
    id: 2,
    title: 'Item title',
    body: 'Item body',
  },
  {
    id: 3,
    title: 'Item title',
    body: 'Item body',
  },
  {
    id: 4,
    title: 'Item title',
    body: 'Item body',
  },
];
const buildAccordion = () => (
  <Accordion>
    {items.map(item => (
      <Accordion.Item key={item.id} onOpen={() => {}} eventKey={item} id={item.id}>
        <Accordion.Item.Title>{item.title + item.id}</Accordion.Item.Title>
        <Accordion.Item.Body>{item.body + item.id}</Accordion.Item.Body>
      </Accordion.Item>
    ))}
  </Accordion>
);

storiesOf('Accordion', module).add('Default accordion', buildAccordion).addDecorator(jsxDecorator);
