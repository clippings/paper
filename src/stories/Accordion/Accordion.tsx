export const items = [
  {
    id: 1,
    title: 'Item title',
    body: 'Item body',
  },
  {
    id: 2,
    title: 'Item title',
    body: 'Item body',
  },
  {
    id: 3,
    title: 'Item title',
    body: 'Item body',
  },
  {
    id: 4,
    title: 'Item title',
    body: 'Item body',
  },
];
