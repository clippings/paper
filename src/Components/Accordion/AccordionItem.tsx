import React from 'react';
import classnames from 'classnames';
import AccordionTitle from './AccordionTitle';
import AccordionBody from './AccordionBody';
import { createClassNameModifier } from '../../utils/classNames';
const classNames = {
  item: createClassNameModifier('accordion', 'item'),
};
type AccordionItemProps = {
  onItemClick?: (...args: any[]) => any;
  onOpen?: (...args: any[]) => any;
  eventKey?: number | any | string;
  active?: boolean;
  index?: number;
};

export type AccordionItemType = React.FunctionComponent<AccordionItemProps>;
const AccordionItem: AccordionItemType = ({
  children,
  onItemClick,
  onOpen,
  active,
  index,
  eventKey,
}) => {
  const handleOpen = () => {
    onItemClick && onItemClick(index);
    onOpen && onOpen(eventKey);
  };
  const title = React.Children.toArray(children).filter(child => child.type === AccordionTitle)[0];
  const body = React.Children.toArray(children).filter(child => child.type === AccordionBody)[0];
  return (
    <div className={classnames(classNames.item, { collapsed: !active })}>
      <AccordionTitle {...title.props} onClick={handleOpen} />
      <AccordionBody {...body.props} />
    </div>
  );
};
AccordionItem.defaultProps = {
  active: false,
  index: 0,
  eventKey: null,
};
AccordionItem.Title = AccordionTitle;
AccordionItem.Body = AccordionBody;
export default AccordionItem;
