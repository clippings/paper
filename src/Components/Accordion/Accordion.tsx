import React, { useState } from 'react';
import classnames from 'classnames';
import { createClassName } from '../../utils/classNames';
import AccordionItem, { AccordionItemType } from './AccordionItem';
import '../../assets/scss/accordion.scss';

const classNames = {
  container: createClassName('accordion'),
};
type AccordionProps = {
  defaultOpened?: number;
};
type AccordionType = React.FunctionComponent<AccordionProps> & { Item: AccordionItemType };

export const Accordion: AccordionType = ({ children, defaultOpened }) => {
  const [active, setActive] = useState(defaultOpened);
  const onItemClick = index => setActive(active === index ? null : index);
  return (
    <div className={classnames(classNames.container)}>
      {React.Children.toArray(children).map((item, index) => {
        const isActive = active === index;
        return (
          <Accordion.Item
            {...item.props}
            active={isActive}
            onItemClick={onItemClick}
            index={index}
            key={`accordion-${index + 1}`}
          />
        );
      })}
    </div>
  );
};

Accordion.Item = AccordionItem;
