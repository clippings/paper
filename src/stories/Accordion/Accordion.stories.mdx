import { Meta, Story, ArgsTable, Canvas } from '@storybook/addon-docs/blocks';
import { Accordion } from '@paper/components';
import { items } from './Accordion';

<Meta title="Accordion" />

export const Template = args => (
  <Accordion defaultOpened={null} disabled={false} {...args} >
    {items.map(item => (
      <Accordion.Item key={item.id} onOpen={() => {}} eventKey={item} id={item.id}>
        <Accordion.Item.Title>{item.title + item.id}</Accordion.Item.Title>
        <Accordion.Item.Body>{item.body + item.id}</Accordion.Item.Body>
      </Accordion.Item>
    ))}
  </Accordion>
);


# Accordion
Basic Accordion component usage.

#### **Accordion**
<ArgsTable of={Accordion}/>

#### **Accordion.Item**
<ArgsTable of={Accordion.Item}/>

#### **Accordion.Item.Title**
<ArgsTable of={Accordion.Item.Title}/>

#### **Accordion.Item.Body**
<ArgsTable of={Accordion.Item.Body}/>

<Canvas>
  <Story
    name="Accordion"
    args={{}}
  >
    {Template.bind({})}
  </Story>
</Canvas>


