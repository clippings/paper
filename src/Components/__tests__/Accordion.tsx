import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Accordion } from '../Accordion/Accordion';
const onOpen = jest.fn();
const accordion = (
  <Accordion>
    <Accordion.Item key="test" onOpen={onOpen} eventKey="test" id={1}>
      <Accordion.Item.Title>Test</Accordion.Item.Title>
      <Accordion.Item.Body>Test content</Accordion.Item.Body>
    </Accordion.Item>
    <Accordion.Item key="test-2" onOpen={onOpen} eventKey="test-2" id={2}>
      <Accordion.Item.Title>Test 2</Accordion.Item.Title>
      <Accordion.Item.Body>Test content 2</Accordion.Item.Body>
    </Accordion.Item>
  </Accordion>
);
describe('Accordion component', () => {
  test('matches snapshot', () => {
    const { container } = render(accordion);
    expect(container).toMatchSnapshot();
  });
  test('onOpen is triggered', () => {
    const { getByText } = render(accordion);
    const title = getByText('Test');
    fireEvent.click(title);
    expect(onOpen).toBeCalled();
    expect(onOpen).toBeCalledWith('test');
  });
});
