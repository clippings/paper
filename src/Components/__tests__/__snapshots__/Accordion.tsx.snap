// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion component matches snapshot 1`] = `
<div>
  <div
    class="p-accordion"
  >
    <div
      class="p-accordion__item collapsed"
    >
      <div
        class="p-accordion__title"
        role="presentation"
      >
        <div
          class="p-accordion__title__wrap"
        >
          Test
        </div>
        <svg
          fill="#141414"
          height="16"
          preserveAspectRatio="xMidYMid meet"
          style="transform: rotate(90deg);"
          viewBox="0 0 14 14"
          width="16"
        >
          <path
            clip-rule="evenodd"
            d="M11.0607 7L5.53032 12.5303L4.46966 11.4697L8.93933 7L4.46966 2.53033L5.53032 1.46967L11.0607 7Z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div
        class="p-accordion__body"
      >
        Test content
      </div>
    </div>
    <div
      class="p-accordion__item collapsed"
    >
      <div
        class="p-accordion__title"
        role="presentation"
      >
        <div
          class="p-accordion__title__wrap"
        >
          Test 2
        </div>
        <svg
          fill="#141414"
          height="16"
          preserveAspectRatio="xMidYMid meet"
          style="transform: rotate(90deg);"
          viewBox="0 0 14 14"
          width="16"
        >
          <path
            clip-rule="evenodd"
            d="M11.0607 7L5.53032 12.5303L4.46966 11.4697L8.93933 7L4.46966 2.53033L5.53032 1.46967L11.0607 7Z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div
        class="p-accordion__body"
      >
        Test content 2
      </div>
    </div>
  </div>
</div>
`;
